coap-gateway/service/observation/resourcesObserver.go

```diff
@@ -0,0 +1,338 @@
+		resources: make(map[int64]*observedResource),
+		coapConn:  coapConn,
+		callbacks: callbacks,
+	}
+}
+
+// Add resource to observer with given interface and wait for initialization message.
+func (o *resourcesObserver) addResource(ctx context.Context, res *commands.Resource, obsInterface string) error {
+	o.lock.Lock()
+	defer o.lock.Unlock()
+	err := o.addResourceLocked(ctx, res, obsInterface)
+	return err
+}
+
+func (o *resourcesObserver) addResourceLocked(ctx context.Context, res *commands.Resource, obsInterface string) error {
+	resID := res.GetResourceID()
+	log.Debugf("observation of resource(%v) requested", resID)
+	addObservationError := func(err error) error {
+		return fmt.Errorf("cannot add resource observation: %w", err)
+	}
+	if o.deviceID == "" {
+		return addObservationError(emptyDeviceIDError())
+	}
+	instanceID := resource.GetInstanceID(resID.GetHref())
+	if o.deviceID != resID.GetDeviceId() {
+		return addObservationError(fmt.Errorf("invalid deviceID(%v)", resID.GetDeviceId()))
+	}
+	href := resID.GetHref()
+	if _, ok := o.resources[instanceID]; ok {
+		return nil
+	}
+	obsRes := NewObservedResource(href, obsInterface)
+	if err := o.handleResourceLocked(ctx, obsRes, res.IsObservable()); err != nil {
+		return addObservationError(err)
+	}
+	o.resources[instanceID] = obsRes
+	return nil
+}
+
+// Handle given resource.
+//
+// For observable resources subscribe to observations, for unobservable resources retrieve
+// their content.
+func (o *resourcesObserver) handleResourceLocked(ctx context.Context, obsRes *observedResource, isObservable bool) error {
+	if obsRes.Href() == commands.StatusHref {
+		log.Debugf("observation of resource /%v%v skipped", o.deviceID, obsRes.Href())
+		return nil
+	}
+
+	if isObservable {
+		obs, err := o.observeResourceLocked(ctx, obsRes)
+		if err != nil {
+			return err
+		}
+		obsRes.SetObservation(obs)
+		return nil
+	}
+	return o.getResourceContentLocked(ctx, obsRes.Href())
+}
+
+// Register to COAP-GW resource observation for given resource
+func (o *resourcesObserver) observeResourceLocked(ctx context.Context, obsRes *observedResource) (*tcp.Observation, error) {
+	cannotObserveResourceError := func(deviceID, href string, err error) error {
+		return fmt.Errorf("cannot observe resource /%v%v: %w", deviceID, href, err)
+	}
+	if o.deviceID == "" {
+		return nil, cannotObserveResourceError(o.deviceID, obsRes.Href(), emptyDeviceIDError())
+	}
+
+	var opts []message.Option
+	if obsRes.Interface() != "" {
+		opts = append(opts, message.Option{
+			ID:    message.URIQuery,
+			Value: []byte("if=" + obsRes.Interface()),
+		})
+	}
+
+	initialized := false
+	obs, err := o.coapConn.Observe(ctx, obsRes.Href(), func(msg *pool.Message) {
+		if !initialized {
+			initialized = true
+			observable := true
+			if _, errObs := msg.Observe(); errObs != nil {
+				log.Debugf("href: %v not observable err: %v", obsRes.Href(), errObs)
+				observable = false
+			}
+			if !observable {
+				return
+			}
+		}
```

Comment:
Yeah, I wanted to ask what to do here. `Observe` sends `Get` for non-observable resource. For non-observable `/oic/res` I want this to do nothing, no notification should be send and the observation as a whole will fall back to observing of published resources. And for published resources I should know whether the resource is observable or not. Only for observable resources will `coapConn.Observe` be called, non-observable call `coapConn.Get`. 

So if the detection of observability is reliable then this will work correctly. But if we want to be defensive then we can do it that for batch `/oic/res` it will do this and in other cases it will call `callbacks.OnObserveResource` (or maybe it should call `callbacks.OnGetResourceContent`? Since for non-observable resource it gets a `Get` response.)

Reply:
changed so this happens only for `/oic/res`